docs/disks.md

```diff
@@ -0,0 +1,41 @@
+# RAID setup
+
+## Erasing existing metadata
+
+If your disks have been used in an array before, the metadata will need wiped before reusing them. **This will destroy the data on the disk**.
+
+```
+$ mdadm --misc --zero-superblock /dev/disk1
+```
+
+## Creating an array
+
+```
+mdadm --create \
+      --level=1 \
+      --metadata=0.90 \
+      --raid-devices=2 \
+      /dev/md/balena-raid \
+      /dev/disk1 \
+      /dev/disk2
+```
+
+Note that an MD device node will be automatically created/acquired by mdadm, and the path give under `/dev/md` will be symlinked to the device node.
+
+Substitute any other RAID level as required. Metadata 0.90 is [required](https://raid.wiki.kernel.org/index.php/RAID_superblock_formats#A_Note_about_kernel_autodetection_of_different_superblock_formats) for kernel auto-detection and automatic assembly of arrays. This is currently required to boot balenaOS.
+
```

Comment:
@jakogut so the customer need to use a different OS first to create/assemble the array? Why can't balenaOS perfrom these steps, is there using input of some sort?

Reply:
@alexgg You can definitely use balenaOS, or any other Linux-based OS that has mdadm. You can't do an online conversion of an existing system to use RAID, however.

The flasher could perform array creation, but it would have to be configured somehow. RAID is an advanced level feature, I would expect anybody using it would be capable of following a few simple steps to create the array.

Reply:
thanks @jakogut so to limit as much as possible the user's friction when using RAID, could this be automated in the initramfs/flasher?

Reply:
@alexgg Once the OS is flashed to a non-MD formatted disk, we're kind of stuck. I believe the disk could be formatted for software RAID from the initramfs before mounting the rootfs, but it would result in losing the contents of the disk.

It is possible that we could have a mechanism to inject a configuration into the flasher to automatically create an array, but I could see that becoming rather complex. There's also greater chances for data loss. A flasher image configured to format all drives in a system as part of a RAID 1 array would erase all the disks in a machine, unprompted, if accidentally booted from.

With networking, we also have a lot of configuration options, but with the benefit that there's a standard configuration file format, and we can just copy it from the boot partition into the rootfs, and NetworkManager will pick it up. Mdadm has a [configuration file](https://man7.org/linux/man-pages/man5/mdadm.conf.5.html) that handles how arrays are assembled, but not how they're created. Probably because doing it from the command line is relatively simple and well documented. Faulty or misapplied network configurations also don't destroy data.

I'm open to ideas for how to improve the process, but at the moment, I think the simplest method is just boot a non-flasher balenaOS image (or other live Linux image of your choosing), manually create the array(s), and write balenaOS to your boot drive. The only part that's vaguely balenaOS specific is not auto detecting and assembling arrays with metadata version greater than 0.90 (probably fixable with a udev hook), everything else is standard Linux tooling.

Reply:
hi @jakogut 

>  but I could see that becoming rather complex.

I really think we should flesh this out in the improvement. We can probably come up with a simple configuration that caters for 80% of cases, and leave the rest for manual creation. RAID configuration is something that Linux installers do, so probably taking a look at some, like the Fedora installer for example, would be useful. @mtoman can probably help out here.

> A flasher image configured to format all drives in a system

Not sure I follow - booting from a flasher image will always erase all data, how is RAID making a difference here?

Reply:
> We can probably come up with a simple configuration that caters for 80% of cases, and leave the rest for manual creation.

@alexgg I'm not sure there's any situation where an automatic RAID configuration covers a majority, or even a sizeable minority of use cases. It's necessarily configured on an application specific basis, as the tradeoffs between read and write performance, fault tolerance, and minimum number of disks vary quite a bit between levels and configurations.

Looking at the Fedora installation wizard, it will create a RAID array out of the disks that are selected to be part of the installation, but the configuration is manual. It's also interactive, so we can see and select disks that are available at runtime, see existing partitions and data, and select the desired array configuration.

![Screenshot_20220822_130947](https://user-images.githubusercontent.com/40163/186009945-f752f822-afd6-433f-b12a-02c75a118539.png)

If I choose to inject a static configuration into a flasher image to automatically provision a software RAID array, there's no guarantee that the disks I expect to be present at runtime will be. I might have eMMC, an SD card, and a thumb drive or portable hard drive plugged in. Which disks are added to the array?

> booting from a flasher image will always erase all data

I don't like how this _currently_ works. I still think this is a bad idea, as even accidentally booting from a balenaOS flasher will erase your disk and destroy your data unprompted. Some PCs might even boot from USB by default, so even writing a flasher image to a thumb drive, forgetting about it, and leaving it plugged in while rebooting could erase data. No other tool in common use that I know of defaults to destroying data unprompted. The difference with automatic RAID configuration is that not just _one_ of my disks would be erased, but _all_ of the disks in the machine. I think this is the wrong direction to move in.

I think if we really wanted to support a feature like this, an interactive installer like what Fedora has might be a safer, more user friendly option.

Reply:
> even accidentally booting from a balenaOS flasher will erase your disk and destroy your data unprompted

Here's another example of why this is so dangerous. I don't use Windows anymore, but I remember that it will reboot to apply updates after a countdown without the user canceling. What if a user has their machine configured to boot from USB first (might be convenient for somebody that boots a lot of Live distros), they start Etcher writing a `generic x86-64` balenaOS flasher to a thumb drive, and they walk away to take lunch. They come back to find that Windows rebooted their machine automatically after the flash completed, and all their data is gone. It would be even worse if the image was configured to setup RAID, as now not only the first disk is erased, but _all_ their disks are erased.

Reply:
Fedora's installer, [Anaconda](https://github.com/rhinstaller/anaconda) seems pretty user friendly, is GPL-2.0 licensed, and is also used by other Linux distros.

Reply:
@jakogut maybe we should actually take this discussion to the improvement, but let me quickly rebate the two main points:

> If I choose to inject a static configuration into a flasher image to automatically provision a software RAID array

We are not speaking about a static configuration, we are speaking about injecting configuration with a graphical interface in the advanced section of the image download dashboard UI.

> booting from a flasher image will always erase all data

I don't think this should be part of this discussion. This is how the flasher works now and RAID does not change this. A different approach to flasher can be discussed of course in a product call, but as I said it RAID changes nothing.

Reply:
This was discussed in a product call and decided we will not extend balenaOS with RAID setup capabilities and leave that to the customer/manufacturing.